### PR TITLE
core.scm:local_url=allow|block

### LINKED ISSUE #16591: [bug] Recipes with invalid remote URLs in conandata can be uploaded
### Describe the bug

Conan 1.64.1

### How to reproduce it

When using the `export()`/`source()` methods in combination with manually implemented url and commit capturing (as recommended in the docs as replacement for the deprecated `scm` attribute) there is one code path which can result in Conan uploading invalid recipes to a remote. Let's suppose we have the following export method:

```python
def export(self):
    git = Git(self, self.recipe_folder)
    scm_url, scm_commit = git.get_url_and_commit()
    update_conandata(self, {"sources": {"commit": scm_commit, "url": scm_url}})
```

The `get_url_and_commit()` method will fail as expected when there are uncommitted changes on the local working copy. It will warn when the local commit is not in the remote (but still return a valid url), which is already problematic. But the actual critical code path is when either the local working copy has no remote set (which can happen in some Gitlab CI merge request pipelines where a separate working copy is created that contains only a single detached ref) or the remote is temporarily unreachable. In both of these cases Conan will silently set the url to the path of the working copy in the local filesystem. This path gets written into the conandata.yml and Conan will happily upload that to the remotes afterwards.

There are several reasons why I think this is a bug:
- Conanfile implementors must manually handle this in the `export()` methods, causing duplicate code in all recipe or alternatively a mandatory python_requires with the shared code
- It is hard to determine the error state because the method encodes this state into the url return value instead of adding a separate boolean that indicates the error, so we'd have to parse the url and decide if it's a valid url or a filesystem path
- Intercepting this code path globally is not possible without having to mock-patch the `get_url_and_commit()` method from a `pre_export()` hook (see below) because the conandata format is not standardized, so we can not deterministically parse it in a `post_export()` hook or the pipelines since we don't know how the url gets stored by the recipe
- The old `scm` attribute prevented upload of recipes with invalid urls, so I consider this a regression

This was breaking our release builds due to temporary Git remote connection issues, leading to broken releases in the remotes. Here is the hook that I'm currently using to mitigate the problem on our side, but I'd rather not put that into production because it could break any day:

```python
from urllib.parse import urlparse
from unittest.mock import patch
from conan.errors import ConanException
from conan.tools.scm import Git


original_get_url_and_commit = Git.get_url_and_commit

def custom_get_url_and_commit(self):
    url, commit = original_get_url_and_commit(self)
    try:
        url_parsed = urlparse(url)
        if url_parsed.scheme not in ["http", "https", "ssh"]:
            raise ValueError()
        return url, commit
    except ValueError:
        raise ConanException("Invalid URL returned by get_url_and_commit(). This might be caused by a missing remote on the working copy or temporary Git remote connection issues.")

def pre_export(output, conanfile, conanfile_path, reference, *args, **kwargs):
    global patcher
    patcher = patch.object(Git, 'get_url_and_commit', new=custom_get_url_and_commit)
    global mock_method
    mock_method = patcher.start()
```

I need some advice on how to proceed here without having to mock-patch or creating our own Conan fork.

### PR DESCRIPTION
**Pull request by memsharded**
Changelog: Feature: Introduce ``core.scm:local_url=allow|block`` to control local folder URLs in conandata ``scm``.
Changelog: Fix: Raise an error when trying to upload a package with a local folder URL in ``conandata`` ``scm``. 
Docs: https://github.com/conan-io/docs/pull/3801

Close https://github.com/conan-io/conan/issues/16591

### CODE REVIEW

File: conans/client/cmd/uploader.py [hunk 1]
```diff
@@ -81,10 +81,17 @@ def __init__(self, app: ConanApp, global_conf):
...
+        local_url = self._global_conf.get("core.scm:local_url", choices=["allow", "block"])
         for ref, bundle in upload_bundle.refs().items():
             layout = self._app.cache.recipe_layout(ref)
             conanfile_path = layout.conanfile()
             conanfile = self._app.loader.load_basic(conanfile_path)
+            url = conanfile.conan_data.get("scm", {}).get("url") if conanfile.conan_data else None
+            if local_url != "allow" and url is not None:
+                if not any(url.startswith(v) for v in ("ssh", "git", "http", "file")):
+                    raise ConanException(f"Package {ref} contains conandata scm url={url}\n"
+                                         "This isn't a remote URL, the build wontt be reproducible\n"
```
Comment (AbrilRBS) [inline conans/client/cmd/uploader.py]:
```suggestion
                                         "This isn't a remote URL, the build won't be reproducible\n"
```

**Comment by jasal82** [comment]
This will work if everyone stores the url in `scm.url`. But your migration guide already recommends otherwise for replacing the scm attribute

https://docs.conan.io/1/migrating_to_2.0/recipes.html#the-scm-attribute

So you'd have to include `sources.url` as well. And any other possible structure because it is never validated. I think we'd have to create an interface for interacting with conandata to be able to do this reliably.

**Comment by memsharded** [comment]
> This will work if everyone stores the url in scm.url. But your migration guide already recommends otherwise for replacing the scm attribute

This has been normalized in Conan 2 with the ``coordinates_to_conandata`` and ``checkout_from_conandata_coordinates`` helpers, that will always store it in ``conandata->scm->url``. The intention is to fully normalize it, and even leverage this information for other functionality.


> https://docs.conan.io/1/migrating_to_2.0/recipes.html#the-scm-attribute

This is a mistake in the docs, and should be updated to reflect the new convention.

**Comment by jasal82** [comment]
I see, that's good to know. That means I'll have to create another release of our base conanfile that uses the conventions from those new functions, so that we can replace them with the actual function calls once we're ready to switch to Conan 2.

**Comment by memsharded** [comment]
Yes, sorry about that, the "standardization" of the ``conandata->scm->url/commit`` happened in https://github.com/conan-io/conan/pull/15377,  that was 1 year after the Conan 2.0 launch, so the Conan 1 docs were no longer updated with these things.

**Review by AbrilRBS — APPROVED** [review]

**Review by czoido — APPROVED** [review]

**Review by ErniGH — APPROVED** [review]

**Review by AbrilRBS — APPROVED** [review]

**Pull request merged**